src/treemodel/selection.js

```diff
@@ -0,0 +1,261 @@
+		 *
+		 * @param {treeModel.Range} range Range to add.
+		 * @param {Boolean} [isBackward] Flag describing if added range was selected forward - from start to end (`false`)
+		 * or backward - from end to start (`true`). Defaults to `false`.
+		 */
+		addRange( range, isBackward ) {
+			pushRange.call( this, range );
+			this._lastRangeBackward = !!isBackward;
+
+			this.fire( 'update' );
+		}
+
+		/**
+		 * Unbinds all events previously bound by this selection, including events bound by created {@link treeModel.LiveRange}s.
+		 */
+		detach() {
+			detachRanges.call( this );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#getAttr}
+		 */
+		getAttr( key ) {
+			return this._attrs.getAttr( key );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#getAttrs}
+		 */
+		getAttrs() {
+			return this._attrs.getAttrs();
+		}
+
+		/**
+		 * Returns an array of ranges added to the selection. The method returns a copy of internal array, so
+		 * it will not change when ranges get added or removed from selection.
+		 *
+		 * @returns {Array.<LiveRange>}
+		 */
+		getRanges() {
+			return this._ranges.slice();
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#hasAttr}
+		 */
+		hasAttr( key ) {
+			return this._attrs.hasAttr( key );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#removeAttr}
+		 */
+		removeAttr( key ) {
+			this._attrs.removeAttr( key );
+		}
+
+		/**
+		 * Removes all ranges that were added to the selection. Fires update event.
+		 */
+		removeAllRanges() {
+			detachRanges.call( this );
+			this._ranges = [];
+
+			this.fire( 'update' );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#setAttr}
+		 */
+		setAttr( attr ) {
+			this._attrs.setAttr( attr );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#setAttrsTo}
+		 */
+		setAttrsTo( attrs ) {
+			this._attrs.setAttrsTo( attrs );
+		}
+
+		/**
+		 * Replaces all ranges that were added to the selection with given array of ranges. Last range of the array
+		 * is treated like the last added range and is used to set {@link #anchor} and {@link #focus}. Accepts a flag
+		 * describing in which way the selection is made (see {@link #addRange}).
+		 *
+		 * @param {Array.<treeModel.Range>} newRanges Array of ranges to set.
+		 * @param {Boolean} [isLastBackward] Flag describing if last added range was selected forward - from start to end (`false`)
+		 * or backward - from end to start (`true`). Defaults to `false`.
+		 */
+		setRanges( newRanges, isLastBackward ) {
+			detachRanges.call( this );
+			this._ranges = [];
+
+			for ( let i = 0; i < newRanges.length; i++ ) {
+				pushRange.call( this, newRanges[ i ] );
+			}
+
+			this._lastRangeBackward = !!isLastBackward;
+			this.fire( 'update' );
+		}
+	}
+
+	/**
+	 * Unbinds all events bound by created {@link treeModel.LiveRange}s.
+	 *
+	 * @private
+	 * @method detachRanges
+	 * @memberOf {treeModel.Selection}
+	 */
+	function detachRanges() {
```

Comment:
Hm... since I don't understand this explanation it may be a good idea to move this code to `Selection.detach()` :D. Also, not keeping the logic totally private allows you to change it if you choose to derive from this class. E.g. `setRanges()` uses `detachRanges()` now. If you'd create `Selection`'s subclass and change how the `detach()` works you would not affect how `setRanges()` works even though it seems to require the same logic.

It's theory and most likely won't happen in this case, but it's a general argument for keeping stuff in the prototype.